doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,522 @@
+# Callback taking index of batch element and all results for that element.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+        batch: List[EntityReference], relationship,
+        successCb: SuccessCallback, errorCb):
+
+    cursor = get_my_connection().cursor()
+
+    # Params for query.
+    my_relationship_type: str = get_my_relationship_type(relationship)
+    entity_ref_strs = (str(ref) for ref in batch)
+
+    # Run the query.
+    placeholders = ", ".join(["?"] * len(batch))  # "?, ?, ?, ..."
+    cursor.execute(
+        f"""
+        SELECT FromRef, ToRef FROM Relationship
+        WHERE RelationshipType = ?
+        AND FromRef IN ({placeholders})
+        ORDER BY FromRef
+        """, my_relationship_type, *entity_ref_strs)
+
+    # Get the raw results
+    result_set = cursor.fetchall()
+
+    # Gather the results for each input ref.
+    result_dict = collections.defaultdict(list)
+    for result in result_set:
+        from_ref = EntityReference(result[0])
+        to_ref = EntityReference(result[1])
+        result_dict[from_ref].append(to_ref)
+
+    # Call callbacks.
+    for idx, from_ref in enumerate(batch):
+        successCb(idx, result_dict[from_ref])
+```
+
+#### Page token signature
+
+```python
+class PageToken(AbstractPageToken):
+    def __init__(self, cursor):
+        self.cursor = cursor
+
+
+def relatedEntities(
+    input: EntityReference, relationship, limit=-1, pageToken=None
+) -> Tuple[AbstractPageToken, List[EntityReferences]]:
+
+    if pageToken is None:
+        my_relationship_type: str = get_my_relationship_type(relationship)
+
+        cursor = get_my_connection().cursor()
+        cursor.execute(
+            f"""
+            SELECT ToRef FROM Relationship
+            WHERE RelationshipType = ?
+            AND FromRef = ?
+            """, my_relationship_type, str(input))
+
+        pageToken = PageToken(cursor)
+
+    if limit <= 0:
+        result_set = pageToken.cursor.fetchall()
+    else:
+        result_set = pageToken.cursor.fetchmany(limit)
+
+    related_refs = [EntityReference(result[0]) for result in result_set]
+    return pageToken, related_refs
+```
+
+#### Cursor object signature
+
+The design of DB-API cursors have a somewhat different control flow to
+the `AbstractPage` example used throughout, and in practice we would
+probably like to mirror a standard such as DB-API in any final design.
+However, shimming the DB-API cursor to match our contrived and
+ill-conceived `AbstractPage` is a useful exercise.
+
+```python
+class Page(AbstractPage):
+    def __init__(self, cursor):
+        self.__cursor = cursor
+        self.__is_pending = False
+        self.__page_size = 0
+
+    def get(self) -> List[EntityReference]:
+        self.__is_pending = False
+        result_set = self.__cursor.fetchmany(self.__page_size)
+        return [
+            EntityReference(result) for result[0] in result_set]
+
+    def next(self, pageSize):
```

Comment:
[question] Is this illustrating the additional cursor functionality (ie arbitrarity advancing without fetching) as assuming (its not entirely clear in the docs) `fetchmany` advances the cursor anyway? If so - maybe calling it `skip` or similar may disambiguate the a host needs to call next to get the next page?

Reply:
Yeah, the whole `AbstractPage` design is a bit wonky and ill-conceived.  

I didn't want to go back and change all the previous sketches, and thought actually the fact that it doesn't conform to DP-API cursor objects provided a useful exercise in adapting a "real" cursor to an awkward design.  

Given that we can't predict the API of back-end cursor objects, I thought that was a valid exercise.

Reply:
Oh to actually answer your question, yeah this is illustrating advancing without fetching, and indeed `fetchmany` also advances the cursor.

However, `next` _must_ be called by the user to set the page size for the `get` call, so "skip" doesn't really work, since it's only skipping if `get` isn't called.

Reply:
Yeah - but I think we have to balance confusing a reader of the DR who will consider this as a proposed cursor design.  The one presented would certainly put me off choosing that option.

Reply:
Agreed, will redesign throughout.